aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1263,6 +1289,94 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
   }
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_eig(const Tensor& self, bool eigenvectors, Tensor& out_eigvals, Tensor& out_eigvecs, magma_int_t *info_ptr) {
+#ifndef USE_MAGMA
+AT_ERROR("symeig: MAGMA library not found in "
+    "compilation. Please rebuild with MAGMA.");
+#else
+  magma_vec_t jobvr = eigenvectors ? MagmaVec : MagmaNoVec;
+  magma_int_t n = magma_int_cast(self.size(-1), "n");
+  auto self_data = self.data_ptr<scalar_t>();
+
+  out_eigvals.resize_({2, n});
+  auto out_eigvals_data = out_eigvals.data_ptr<scalar_t>();
+  scalar_t *wr = out_eigvals_data;
+  scalar_t *wi = out_eigvals_data+n;
+
+  scalar_t *vr_data = NULL;
+  magma_int_t ldvr = 1;
+  if (jobvr == MagmaVec)
+  {
+      vr_data = out_eigvecs.data_ptr<scalar_t>();
+      ldvr = n;
+  }
+
+  if (n > 0) {
+    // call magmaEig once to get the optimal size of work_data
+    scalar_t wkopt;
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, &wkopt, -1, info_ptr);
+    magma_int_t lwork = (magma_int_t) wkopt;
+
+    // call it a 2nd time to to the actual work
+    scalar_t *work_data = nullptr;
+    ALLOCATE_ARRAY(work_data, scalar_t, lwork);
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, work_data, lwork, info_ptr);
+  }
+
+  // magmaEig returns column-ordered tensors, turn them into row-ordered
+  out_eigvals.transpose_(0, 1);
+  out_eigvecs.transpose_(0, 1);
+#endif
+}
+
+
+std::tuple<Tensor &,Tensor &> eig_cuda_out(Tensor & e, Tensor & v, const Tensor & self, bool eigenvectors) {
+  Tensor vals_tmp, vecs_tmp;
+  std::tie(vals_tmp, vecs_tmp) = eig_cuda(self, eigenvectors);
+  e.resize_as_(vals_tmp).copy_(vals_tmp);
+  v.resize_as_(vecs_tmp).copy_(vecs_tmp);
```

Comment:
> In general, an important reason for using out argument is to avoid unnecessary memory copies.

I agree; what I did was to follow what it is done by the [existing `symEig`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L922), so I assumed it was "fine".

> IIUC, the pytorch convention of specifying out parameter is as follows [...]

good to know: I'll change the code to do that

Reply:
The problem here (and the reason why it is symeig works how it does) is that vals and vecs might not be contiguous, in which case you cannot expect the backend to deal with them correctly.

Reply:
wouldn't it be possible to check whether `out_vals` and `out_vecs` have the correct shape/format, and fall back to the copy approach only if the don't?

Reply:
Possibly, yes, and it is nice to look at avoiding copies if possible. On the other hand, it isn't entirely clear that it is as crucial for n^3 algorithms as it is in the cheaper functions.
Maybe @vishwakftw has some insight to share from his work on `symeig`.

Reply:
I must admit that altering the behavior of a single function to deviate from the PyTorch convention in favour of some Numpy condition did not occur to me as the ideal course of action for moving from TH to ATen.

Reply:
pytorch and numpy use different conventions in various places and while there exist efforts to unify these conventions, I agree that it makes sense to preserve existing conventions when porting to ATen.

Reply:
so, if I understand it correctly:
- currently, it is supported to pass a non-contiguous tensor, and we expect `.eig` to "fix" it
- in the future, we want to deprecate this behavior and raise an error and/or warning if you pass the "wrong" out vector.

So I propose to already insert the check now:
- if the out vector is "correct", we will use it directly
- if it has not the correct shape/contiguity/whatever, we will use `copy` for now, and raise a warning and/or error in the future.

What do you think?

Reply:
I removed the unnecessary copy in the `out=...` case, see commits ce66134 and 709510f. Now there is a `eig_cuda_helper` function which returns *CPU* tensors (since this is what is returned by `magmaEig` anyway), and `eig_cuda` and `eig_cuda_out` take care of copying the values to the right place.

Moreover, I am using `at::native::resize_output` to handle the output tensors: this way we get a warning in case the shape of the out tensor is not correct. In order to use it, I had to mark it as `C10_EXPORT` in `Resize.h`, because we need to link it from `libtorch_cpu.so` to `libtorch_cuda.so`.

If you agree with the changes, I'm going to mark this conversation as resolved :)

Reply:
Hmm. Are we exporting at::native in general?

Reply:
I don't know what is the official convention, but a quick `grep CAFFE2_API` inside `ATEN/native` shows that there are already exported functions in that namespace:
```
$ grep CAFFE2_API *.h
DispatchStub.h:struct CAFFE2_API DispatchStub;
DispatchStub.h:struct CAFFE2_API DispatchStub<rT (*)(Args...), T> {
DispatchStub.h:  extern CAFFE2_API struct name name
Resize.h:CAFFE2_API void resize_output(Tensor& output, IntArrayRef shape);
TensorIterator.h:struct CAFFE2_API OperandInfo {
TensorIterator.h:struct CAFFE2_API TensorIterator {
TensorIterator.h:class CAFFE2_API TensorIteratorConfig final {
TensorIterator.h:struct CAFFE2_API SplitUntil32Bit {
TensorIterator.h:  struct CAFFE2_API iterator {
TypeProperties.h:CAFFE2_API ResultTypeState update_result_type_state(const Tensor& tensor, const ResultTypeState& in_state);
TypeProperties.h:CAFFE2_API ScalarType result_type(const ResultTypeState& state);
TypeProperties.h:CAFFE2_API ScalarType result_type(TensorList tensors);
```

Reply:
@t-vi can I mark this conversation as resolved, or do you want to add more?